mbq/atomiq/producers.py

```diff
@@ -49,8 +49,36 @@ def _is_running_within_transaction(self):
         save points and expect there at least to be 2: 1 from TestCase and 1 user-defined.
         """
         if RUNNING_TESTS:
-            if len(db_connection.savepoint_ids) < 2:
-                return False
+            import inspect
+            from django.test import TestCase
+            inSetup = False
+            inTestCase = False
+
+            # This loops through the call stack and sets "isSetUp" and "isTestCase"
+            for stack_frame in inspect.stack():
+                frame = stack_frame[0]
+                localVars = frame.f_locals
+                selfVar = localVars.get('self')
```

Comment:
This isn't the right way to check this, you're relying on convention here and hilariously you're relying on _the wrong convention_ (since this really ought to be called `cls`)

Instead just loop over all the vars and check for each of them

Reply:
Not getting how it's the wrong convention, and I'm worried about casting too wide of a net if I just checked _every_ local variable in the whole call stack.

Reply:
Further hilarity, there isn't actually a firmly established convention around class in classmethods and some insane people call it `klass`

Could probably make a D&D alignment chart out of what to call class in classmethods

Reply:
I'm thinking `klass` would be chaotic good but I'm open to counteropinions

Reply:
everything I've ever read says that `cls` is the convention. Not convinced anybody at Q would use anything else.

Reply:
Either way I think you want to loop over the locals

Reply:
aren't you worried about getting false positives? Who knows why some class could be imported somewhere in the call stack. There's a lot of locals.

Reply:
TIL PEP8 actually recommends `cls`, neat

Reply:
> aren't you worried about getting false positives? Who knows why some class could be imported somewhere in the call stack. There's a lot of locals.

I'm not, tests don't tend to import other test classes and if you've got Django TestCase in your production code somewhere you deserve all the pain coming your way

Reply:
I'm not convinced but also not opinionated enough to argue. If you add another comment to change it I'll change it.

Reply:
If I break convention by, say, naming my variables with camel case instead of underscores, should the consequences be:

1. Mild inconvenience to other devs that mess around in this project in the future
2. An incredibly hard-to-track down bug leading me to an arcane bit of hackery in a downstream library that took a convention and instead made it a hard runtime requirement

I'm hearing you on the false positive thing but I think it's theoretical. If it's not, prove it, otherwise write code that isn't designed to infuriate people unnecessarily down the line.

Reply:
Oh wait also why aren't you checking function names to see if you're in one of the set up functions you're expecting? Those aren't conventions, those are the TestCase implementation, you can rely on those

Reply:
Breaking the convention would cause the tests to _not_ fail.
So actually, assuming somebody broke the convention _and_ they forgot the transaction, the consequence would actually be:

3.  Atomiq raises a "no transaction" exception in prod. 

but like I said I'll change it.

Reply:
Everything you just laid out is the worst possible outcome, though. That's exactly what we need to prevent.